certora/specs/ERC721.spec

```diff
@@ -0,0 +1,586 @@
+import "helpers.spec"
+import "methods/IERC721.spec"
+
+methods {
+    // exposed for FV
+    mint(address,uint256)
+    safeMint(address,uint256)
+    safeMint(address,uint256,bytes)
+    burn(uint256)
+
+    tokenExists(uint256) returns (bool) envfree
+    unsafeOwnerOf(uint256) returns (address) envfree
+    unsafeGetApproved(uint256) returns (address) envfree
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Helpers                                                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+// Could be broken in theory, but not in practice
```

Comment:
Remember that 0 is a valid tokenId, so there are technically 2**256 tokenIds ... which is max_uint256+1.

If you mint ALL the existing tokens to a single address, the balance overflows a uint256.

Reply:
> I know other invariants depend on this, but, do you think it's safe to assume it will never break in practice?

Do you expect anyone to call mint 2**256 times ? No contract we provide (including ERC721Consecutive) make that realistic.

> Is there a way of doing this without assuming balanceLimited() (eg. using requireInvariant ownedTotalIsSumOfBalances() instead), or this is the official Certora recommendation?

There is no such thing as official recommendation. and `ownedTotalIsSumOfBalances` just says that `ownedTotal() == sumOfBalances()` ... it doesn't say that both can't be max_uint256

Reply:
> Is it still impossible to prove? I have the feeling it can be implicit in ownedTotalIsSumOfBalances

`ownedTotalIsSumOfBalances` sais that is both value are defined, they are equal. It doesn't says that it does not overflow

Reply:
> If you mint ALL the existing tokens to a single address, the balance overflows a uint256.

Ahh I see, this makes it clear, thanks.